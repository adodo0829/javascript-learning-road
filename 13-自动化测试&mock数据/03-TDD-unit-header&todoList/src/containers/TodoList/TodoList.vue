<template>
  <div>
    <Header @add="addUndoItem" />
    <ul>
      <li v-for="item in undoList" :key="item">{{item}}</li>
    </ul>
  </div>
</template>

<script>
import Header from './components/Header'
export default {
  name: 'TodoList',
  components: {
    Header
  },
  data () {
    return {
      undoList: []
    }
  },
  methods: {
    addUndoItem (inputValue) {
      this.undoList.push(inputValue)
    }
  }
}
</script>

<style scoped lang="stylus">
</style>
