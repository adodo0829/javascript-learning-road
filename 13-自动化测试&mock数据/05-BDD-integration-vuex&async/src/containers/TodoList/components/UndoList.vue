<template>
  <div class="undolist">
    <div class="title">
      正在进行
      <span data-test="count" class="count">{{list.length}}</span>
    </div>
    <ul class="list">
      <li v-for="(item, index) in list"
        :key="index"
        data-test="list-item"
        class="item"
        @click="changeStatus(index)"
      >
        <input
          class="input"
          v-if="item.status === 'input'"
          data-test="input"
          :value="item.value"
          @blur="handleInputBlur"
          @change="(e) => changeItemValue(e.target.value, index)"
        >
        <span v-else>{{item.value}}</span>
        <span data-test="delete-button" @click="() => { handleDelete(index) }" class="delete">-</span>
      </li>
    </ul>
  </div>
</template>

<script>
export default {
  name: 'UndoList',
  props: [ 'list' ],
  data () {
    return {
      //
    }
  },
  methods: {
    handleDelete (index) {
      this.$emit('delete', index)
    },
    changeStatus (index) {
      this.$emit('status', index)
    },
    handleInputBlur () {
      this.$emit('reset')
    },
    // handleInputChange (value, index) {
    //   this.$emit('change', {
    //     value,
    //     index
    //   })
    // },
    changeItemValue (value, index) {
      this.$emit('change', {
        value,
        index
      })
    }
  }
}
</script>

<style scoped lang="stylus">
.undolist {
  width: 600px;
  margin: 0 auto;
}
.title {
  margin: 10px 0;
  line-height: 30px;
  font-size: 24px;
  font-weight: bold;
}
.count {
  margin-top: 5px;
  float: right;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-size: 12px;
  text-align: center;
  background: #e6e6e6;
  border-radius: 10px;
  color: #000;
}
.item {
  margin-bottom: 10px;
  line-height: 32px;
  font-size: 14px;
  background: #fff;
  border-left: 5px solid #629A9A;
  border-radius: 3px;
  text-indent: 10px;
}
.list {
  list-style-type: none;
}
.delete {
  margin-top: 5px;
  margin-right: 10px;
  float: right;
  display: block;
  width: 20px;
  height: 20px;
  line-height: 20px;
  font-size: 12px;
  text-align: center;
  background: #e6e6e6;
  border-radius: 10px;
  color: #000;
  text-indent: 0;
  cursor: pointer;
}
.input {
  width: 460px;
  height: 22px;
  text-indent: 10px;
}
</style>
