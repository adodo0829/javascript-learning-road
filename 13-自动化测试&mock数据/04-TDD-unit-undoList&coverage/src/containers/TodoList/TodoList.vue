<template>
  <div>
    <Header @add="addUndoItem" />
    <UndoList
      :list="undoList"
      @delete="handleItemDelete"
      @status="changeStatus"
      @reset="resetStatus"
      @change="changeItemValue"
    />
  </div>
</template>

<script>
import Header from './components/Header'
import UndoList from './components/UndoList'
export default {
  name: 'TodoList',
  components: {
    Header,
    UndoList
  },
  data () {
    return {
      undoList: [
        // {
        //   status: 'input',
        //   value: 1
        // }, {
        //   status: 'input',
        //   value: 1
        // }
      ]
    }
  },
  methods: {
    addUndoItem (inputValue) {
      this.undoList.push({
        status: 'div',
        value: inputValue
      })
    },
    handleItemDelete (index) {
      this.undoList.splice(index, 1)
    },
    changeStatus (index) {
      const newList = []
      this.undoList.forEach((item, itemIndex) => {
        if (itemIndex === index) {
          newList.push({
            status: 'input',
            value: item.value
          })
        } else {
          newList.push({
            status: 'div',
            value: item.value
          })
        }
      })
      this.undoList = newList
    },
    resetStatus () {
      console.log(111)
      let newList = []
      this.undoList.forEach((item, itemIndex) => {
        newList.push({
          status: 'div',
          value: item.value
        })
      })
      this.undoList = newList
    },
    changeItemValue (obj) {
      this.undoList[obj.index].value = obj.value
      // const newList = []
      // this.undoList.forEach((item, itemIndex) => {
      //   if (itemIndex === index) {
      //     newList.push({
      //       status: 'input',
      //       value: item.value
      //     })
      //   } else {
      //     newList.push({
      //       status: 'div',
      //       value: item.value
      //     })
      //   }
      // })
      // this.undoList = newList
    }
  }
}
</script>

<style scoped lang="stylus">
</style>
