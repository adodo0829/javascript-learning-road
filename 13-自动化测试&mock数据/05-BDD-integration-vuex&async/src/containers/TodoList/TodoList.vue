<template>
  <div>
    <Header @add="addUndoItem" />
    <UndoList
      :list="undoList"
      @delete="handleItemDelete"
      @status="changeStatus"
      @reset="resetStatus"
      @change="changeItemValue"
    />
  </div>
</template>

<script>
import axios from 'axios'
import Header from './components/Header'
import UndoList from './components/UndoList'
export default {
  name: 'TodoList',
  components: {
    Header,
    UndoList
  },
  data () {
    return {
      undoList: [
        // {
        //   status: 'input',
        //   value: 1
        // }, {
        //   status: 'input',
        //   value: 1
        // }
      ]
    }
  },
  mounted () {
    /**
     * {
     *    success: true,
     *    data: [{
     *      status: 'div',
     *      value: 'csxiaoyao'
     *    }]
     * }
     */
    axios.get('/getUndoList.json').then((res) => {
      this.undoList = res.data
    }).catch(e => {
      console.log(e)
    })
    /*
    setTimeout(() => {
      axios.get('/getUndoList.json').then((res) => {
        this.undoList = res.data
      }).catch(e => {
        console.log(e)
      })
    }, 3000)
    */
  },
  methods: {
    addUndoItem (inputValue) {
      this.undoList.push({
        status: 'div',
        value: inputValue
      })
    },
    handleItemDelete (index) {
      this.undoList.splice(index, 1)
    },
    changeStatus (index) {
      const newList = []
      this.undoList.forEach((item, itemIndex) => {
        if (itemIndex === index) {
          newList.push({
            status: 'input',
            value: item.value
          })
        } else {
          newList.push({
            status: 'div',
            value: item.value
          })
        }
      })
      this.undoList = newList
    },
    resetStatus () {
      console.log(111)
      let newList = []
      this.undoList.forEach((item, itemIndex) => {
        newList.push({
          status: 'div',
          value: item.value
        })
      })
      this.undoList = newList
    },
    changeItemValue (obj) {
      this.undoList[obj.index].value = obj.value
      // const newList = []
      // this.undoList.forEach((item, itemIndex) => {
      //   if (itemIndex === index) {
      //     newList.push({
      //       status: 'input',
      //       value: item.value
      //     })
      //   } else {
      //     newList.push({
      //       status: 'div',
      //       value: item.value
      //     })
      //   }
      // })
      // this.undoList = newList
    }
  }
}
</script>

<style scoped lang="stylus">
</style>
