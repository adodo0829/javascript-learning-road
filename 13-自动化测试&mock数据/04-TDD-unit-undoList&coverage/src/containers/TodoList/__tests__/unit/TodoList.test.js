import { shallowMount } from '@vue/test-utils'
import TodoList from '@/containers/TodoList/TodoList'
import UndoList from '@/containers/TodoList/components/UndoList'

describe('TodoList 组件', () => {
  it('TodoItem 初始化时，undoList 应该为空', () => {
    const wrapper = shallowMount(TodoList)
    const undoList = wrapper.vm.$data.undoList
    expect(undoList).toEqual([])
  })

  it('TodoItem 执行 addItem 时会增加一个内容', () => {
    const wrapper = shallowMount(TodoList)
    wrapper.vm.addUndoItem('csxiaoyao')
    const undoList = wrapper.vm.$data.undoList
    expect(undoList).toEqual([{
      status: 'div',
      value: 'csxiaoyao'
    }])
  })

  it('TodoList 中 addUndoItem 被执行后，内容会增加一项', () => {
    // const content = 'csxiaoyao'
    // const wrapper = shallowMount(TodoList)
    // const header = wrapper.find(Header)
    // header.vm.$emit('add', content)
    // const undoList = wrapper.vm.$data.undoList
    // expect(undoList).toEqual([content])

    const wrapper = shallowMount(TodoList)
    wrapper.setData({
      undoList: [
        { status: 'div', value: 1 },
        { status: 'div', value: 2 },
        { status: 'div', value: 3 }
      ]
    })
    wrapper.vm.addUndoItem(4)
    expect(wrapper.vm.$data.undoList).toEqual([
      { status: 'div', value: 1 },
      { status: 'div', value: 2 },
      { status: 'div', value: 3 },
      { status: 'div', value: 4 }
    ])
  })

  it('UndoList 调用 UndoList，应该传递 list 参数', () => {
    const wrapper = shallowMount(TodoList)
    const undoList = wrapper.find(UndoList)
    const list = undoList.props('list')
    expect(list).toBeTruthy()
  })

  it('changeStatus 方法被调用时，UndoList 列表内容变化', () => {
    const wrapper = shallowMount(TodoList)
    wrapper.setData({
      undoList: [
        { status: 'div', value: 1 },
        { status: 'div', value: 2 },
        { status: 'div', value: 3 }
      ]
    })
    wrapper.vm.changeStatus(1)
    expect(wrapper.vm.$data.undoList).toEqual([
      { status: 'div', value: 1 },
      { status: 'input', value: 2 },
      { status: 'div', value: 3 }
    ])
  })

  it('resetStatus 方法执行时，UndoList 列表内容变化', () => {
    const wrapper = shallowMount(TodoList)
    wrapper.setData({
      undoList: [
        { status: 'div', value: 1 },
        { status: 'input', value: 2 },
        { status: 'div', value: 3 }
      ]
    })
    wrapper.vm.resetStatus(1)
    expect(wrapper.vm.$data.undoList).toEqual([
      { status: 'div', value: 1 },
      { status: 'div', value: 2 },
      { status: 'div', value: 3 }
    ])
  })

  it('changeItemValue 方法执行时，UndoList 列表内容变化', () => {
    const wrapper = shallowMount(TodoList)
    wrapper.setData({
      undoList: [
        { status: 'div', value: 1 },
        { status: 'input', value: 2 },
        { status: 'div', value: 3 }
      ]
    })
    wrapper.vm.changeItemValue({
      value: '444',
      index: 1
    })
    expect(wrapper.vm.$data.undoList).toEqual([
      { status: 'div', value: 1 },
      { status: 'input', value: '444' },
      { status: 'div', value: 3 }
    ])
  })
})
