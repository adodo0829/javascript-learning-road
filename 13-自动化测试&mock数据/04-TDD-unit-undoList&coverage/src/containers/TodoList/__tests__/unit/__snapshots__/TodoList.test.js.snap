// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList.vue 组件渲染正常 1`] = `
<div>
  <h1>TodoList</h1>
</div>
`;
